### ISSUE TITLE
New release for SQLAlchemy >= 2.0.22

### ISSUE THREAD

**Issue by hashworks** [kvesteri/sqlalchemy-utils#729]
#725 added some fixes for SQLAlchemy >= 2.0.22, but there is no new version yet. Could you release one? This is blocking a `python-sqlalchemy` update on Arch Linux.

**Comment by josecsotomorales** [comment]
We are on the same boat here, I see some tests failing but unrelated to the PR merged.

**Comment by josecsotomorales** [comment]
It's failing since this [PR](https://github.com/kvesteri/sqlalchemy-utils/commit/fbefeefa5e96adf82c9f8f63bbe6e6e2c846d187)

**Comment by marksteward** [comment]
Sorry, I completely missed this! I've released 1.4.1, which pins to SQLAlchemy 2.0.21 for now (see https://github.com/sqlalchemy/sqlalchemy/discussions/10474)

**Comment by marksteward** [comment]
This is also breaking sqlalchemy-continuum. For now I've pinned SQLAlchemy but I'm looking at removing the dependency.

**Comment by kvesteri** [comment]
@josecsotomorales I added you as a collaborate, if you send me your PyPI username I can you add there as a maintainer as well.

**Comment by josecsotomorales** [comment]
Thank you @kvesteri, my PyPi username is `josecsoto91`

**Comment by josecsotomorales** [comment]
FYI working on getting `master` to pass tests in this [PR](https://github.com/kvesteri/sqlalchemy-utils/pull/731)

**Comment by josecsotomorales** [comment]
Can I get some help reviewing this [PR](https://github.com/kvesteri/sqlalchemy-utils/pull/732)? There's an assertion that I had to comment, not clear why it's failing.

**Comment by marksteward** [comment]
I don't see how the changes you've made to aggregates.py can be right. What have you done to debug it?

**Comment by josecsotomorales** [comment]
I tested with setting up echo to true and debug the generated queries, with the previous local condition it was generating a cartesian product in the UPDATE ... FROM query from the association table

**Comment by josecsotomorales** [comment]
@kvesteri could you please add @marksteward as a contributor as well?

**Comment by marksteward** [comment]
The test first fails on 2.0.13, presumably due to https://github.com/sqlalchemy/sqlalchemy/commit/4a62625. The query is:

```
UPDATE "user" SET group_count=(SELECT count(:count_2) AS count_1 
FROM "group" JOIN user_group AS user_group_1 ON "group".id = user_group_1.group_id 
WHERE "user".id = user_group_1.user_id) FROM "group" WHERE "group".id IN (__[POSTCOMPILE_id_1])'
```

Which looks wrong to me, it should be updating the `user` table only. It's explicitly pulling out the remote side of the relationship [when a secondary is set](https://github.com/kvesteri/sqlalchemy-utils/blob/master/sqlalchemy_utils/aggregates.py#L410-L412). This code was introduced in 2014 (3ca0e9a), and trying with 0.9 (October 2014), `local_remote_pairs` is the same structure as now (i.e. local pairs to begin, remote pairs following):

```
[(Column('id', Integer(), table=<user>, primary_key=True, nullable=False),
  Column('user_id', Integer(), ForeignKey('user.id'), table=<user_group>)),
 (Column('id', Integer(), table=<group>, primary_key=True, nullable=False),
  Column('group_id', Integer(), ForeignKey('group.id'), table=<user_group>))]
```

The whole of `local_condition` is pretty odd, I don't see how it can work with composite foreign keys, and this code in the other branch is straightforwardly wrong:

```
            remote_pairs = property_.local_remote_pairs
            local = remote_pairs[0][0]
            remote = remote_pairs[0][1]
```

I'd be tempted to just skip the test or drop this code entirely.

**Comment by josecsotomorales** [comment]
Doh! That makes sense, I will rollback and just skip the test for now.

**Comment by josecsotomorales** [comment]
Merged to `master` the config fix by skipping the two problematic tests, from comments in the SA 2.0.22 issue seems like we need to refactor this [PR](https://github.com/kvesteri/sqlalchemy-utils/commit/00d800063d2f8d2dcee714a3314311d79f8d4040)

**Comment by josecsotomorales** [comment]
@marksteward created a [PR](https://github.com/kvesteri/sqlalchemy-utils/pull/733) for SA >= 2.0.22, can you review it? I think this approach will work for older versions as well.

**Comment by marksteward** [comment]
lgtm!

**Comment by josecsotomorales** [comment]
@kvesteri could you please send me another invitation to PyPI? Sorry I missed the previous one.

**Comment by kvesteri** [comment]
@josecsotomorales I invited you again and PyPi and invited @marksteward as collaborator here in github for this project and also as a maintainer for this package in PyPi

**Comment by josecsotomorales** [comment]
Thanks @kvesteri!! Released `0.41.2` on PyPI

**Comment by josecsotomorales** [comment]
Closing this issue now, thanks all 🚀
